File: CloudProgram/cloudfunctions/delete-member/Member.ts
export class Member {
  id: string = '';
  name: string = '';
  studentNo: string = '';
  department: string = '';
  roleId: string = '';
  roleLabel: string = '';
  phone: string = '';
  email: string = '';
  joinedAt: Date = new Date();
  dingTalkUserId: string = '';
  syncStatus: string = '';
  lastSyncedAt: Date | null = null;
  orgId: string = '';
  updatedAt: Date | null = null;

  getClassName(): string {
    return 'Member';
  }

  getFieldTypeMap(): Map<string, string> {
    const map = new Map<string, string>();
    map.set('id', 'String');
    map.set('name', 'String');
    map.set('studentNo', 'String');
    map.set('department', 'String');
    map.set('roleId', 'String');
    map.set('roleLabel', 'String');
    map.set('phone', 'String');
    map.set('email', 'String');
    map.set('joinedAt', 'Date');
    map.set('dingTalkUserId', 'String');
    map.set('syncStatus', 'String');
    map.set('lastSyncedAt', 'Date');
    map.set('orgId', 'String');
    map.set('updatedAt', 'Date');
    return map;
  }

  getPrimaryKeyList(): string[] {
    return ['id'];
  }

  getIndexList(): string[] {
    return ['id'];
  }

  getEncryptedFieldList(): string[] {
    return [];
  }

  static parseFrom(data: any): Member {
    const obj = new Member();
    if (data) {
      obj.id = data.id ?? '';
      obj.name = data.name ?? '';
      obj.studentNo = data.studentNo ?? '';
      obj.department = data.department ?? '';
      obj.roleId = data.roleId ?? '';
      obj.roleLabel = data.roleLabel ?? '';
      obj.phone = data.phone ?? '';
      obj.email = data.email ?? '';
      obj.joinedAt = data.joinedAt ? new Date(data.joinedAt) : new Date();
      obj.dingTalkUserId = data.dingTalkUserId ?? '';
      obj.syncStatus = data.syncStatus ?? '';
      obj.lastSyncedAt = data.lastSyncedAt ? new Date(data.lastSyncedAt) : null;
      obj.orgId = data.orgId ?? '';
      obj.updatedAt = data.updatedAt ? new Date(data.updatedAt) : null;
    }
    return obj;
  }
}

File: Application/lib/models/member.dart
/// 社团成员
class Member {
  Member({
    required this.id,
    required this.name,
    required this.studentNo,
    required this.department,
    required this.roleId,
    required this.roleLabel,
    this.phone = '',
    this.email = '',
    required this.joinedAt,
    this.dingTalkUserId = '',
    this.syncStatus = 'manual',
    this.lastSyncedAt,
  });

  final String id;
  final String name;
  final String studentNo;
  final String department;
  String roleId;
  String roleLabel;
  String phone;
  String email;
  final DateTime joinedAt;
  final String dingTalkUserId;
  final String syncStatus;
  final DateTime? lastSyncedAt;

  int get avatarColorIndex => name.hashCode.abs() % 8;

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'studentNo': studentNo,
        'department': department,
        'roleId': roleId,
        'roleLabel': roleLabel,
        'phone': phone,
        'email': email,
        'joinedAt': joinedAt.millisecondsSinceEpoch,
        'dingTalkUserId': dingTalkUserId,
        'syncStatus': syncStatus,
        if (lastSyncedAt != null) 'lastSyncedAt': lastSyncedAt!.millisecondsSinceEpoch,
      };

  factory Member.fromJson(Map<String, dynamic> json) => Member(
        id: json['id'] as String,
        name: json['name'] as String,
        studentNo: json['studentNo'] as String,
        department: json['department'] as String,
        roleId: (json['roleId'] as String?) ?? (json['role'] as String?) ?? 'member',
        roleLabel: (json['roleLabel'] as String?) ?? '成员',
        phone: (json['phone'] as String?) ?? '',
        email: (json['email'] as String?) ?? '',
        joinedAt: DateTime.fromMillisecondsSinceEpoch(
          (json['joinedAt'] as int?) ??
              DateTime.now().millisecondsSinceEpoch,
        ),
        dingTalkUserId: (json['dingTalkUserId'] as String?) ?? '',
        syncStatus: (json['syncStatus'] as String?) ?? 'manual',
        lastSyncedAt: (json['lastSyncedAt'] as int?) != null
            ? DateTime.fromMillisecondsSinceEpoch(json['lastSyncedAt'] as int)
            : null,
      );
}
